src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs

```diff
@@ -84,5 +87,86 @@ internal static ISettings ProcessConfigFile(string configFile)
                 configFileName,
                 machineWideSettings: new XPlatMachineWideSetting());
         }
+
+        /// <summary>
+        /// Get the project or solution file from the given directory.
+        /// </summary>
+        /// <param name="directory">A directory with exactly one project or solution file.</param>
+        /// <returns>A single project or solution file.</returns>
+        /// <exception cref="ArgumentException">Throws an exception if the directory has none or multiple project/solution files.</exception>
+        internal static string GetProjectOrSolutionFileFromDirectory(string directory)
+        {
+            var topLevelFiles = Directory.GetFiles(directory, "*.*", SearchOption.TopDirectoryOnly);
+
+            var solutionFiles = topLevelFiles
+                                    .Where(file => IsSolutionFile(file))
+                                    .ToArray();
+            var projectFiles = topLevelFiles
+                                    .Where(file => IsProjectFile(file))
+                                    .ToArray();
+
+            if (solutionFiles.Length + projectFiles.Length > 1)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.Error_MultipleProjectOrSolutionFilesInDirectory,
+                        directory));
+            }
+
+            if (solutionFiles.Length == 1)
+            {
+                return solutionFiles[0];
+            }
+
+            if (projectFiles.Length == 1)
+            {
+                return projectFiles[0];
+            }
+
+            throw new ArgumentException(
+                string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.Error_NoProjectOrSolutionFilesInDirectory,
+                    directory));
+        }
```

Comment:
This needs formatting and testing (I did not test it). This should achieve the same objective of validating that there is exactly one and only one solution OR project file (not both) and return it, otherwise throw the appropriate error while saving allocations (solution and project file arrays) and having to iterate through the files list twice. Please be sure to test and format the code and fix any bugs (I wrote this in the browser).

```suggestion
    var topLevelFiles = Directory.GetFiles(directory, "*.*", SearchOption.TopDirectoryOnly);
    string candidateFile = null;
    foreach (string file in topLevelFiles)
    {
        if (IsSolutionFile(file) || IsProjectFile(file))
        {
            // We already found a solution or project file and now we found another one.
            if (candidateFile != null)
            {
                throw new ArgumentException(
                        string.Format(
                            CultureInfo.CurrentCulture,
                            Strings.Error_MultipleProjectOrSolutionFilesInDirectory,
                            directory));
            }
        
            candidateFile = file;
        }
    }
    
    // Still being null after visiting all the files means we didn't find a solution or project file.
    if (candidateFile == null)
    {
            throw new ArgumentException(
                string.Format(
                    CultureInfo.CurrentCulture,
                    Strings.Error_NoProjectOrSolutionFilesInDirectory,
                    directory));
    }
    
    return candidateFile;
```

Reply:
Thanks @jebriede! I've changed the implementation now